### PR TITLE
Fix admin localized date formatting

### LINKED ISSUE #27310: Localized date field formatting in admin messing up actual date
### Preconditions (*)
1. Magento `>= 2.3.3-p1`, `2.4-develop`

### Steps to reproduce (*)

1. Have 2 admin users. One on locale `en_US`, one one `nl_NL`.
2. Log in on `nl_NL` admin user.
3. Create a cart price rule, no coupon code needed, just a simple cart price rule. Make sure you set the **from date** (e.g. 17 march 2020). Save the cart price rule.

### Expected result (*)

1. The saved **from date** is `17-03-2020`.
image

### Actual result (*)

1. The saved **from date** is `20-03-2017`.
image

### PR DESCRIPTION
**Pull request by tdgroot**
### Description (*)
There were 2 issues going on that are fixed by this PR. 

1. **Incorrect initial value parsing**
When parsing the initial value, say `2020-03-17`, the date was getting parse with the following format (for nl_NL): `dd-MM-y`. The value used was derived from the variable `outputDateFormat`. I changed the format by using the variable `inputeDateFormat`, which is `yyyy-MM-dd`.
2. **Cyclic event chain**
Variable `shiftedValue` was initialized by the `onValueChange` listener. But when changing `shiftedValue`, the `value` variable is changed by the `onShiftedValueChange` listener. This would mess up the date value as well. 
A cyclic event chain like the following could occur:
    - Change `shiftedValue` to `2020-03-17`.
    - Change `value` to `17-03-2020`.
    - Change `shiftedValue` to `20-03-2017`. 

### Fixed Issues (if relevant)
1. fixes magento/magento2#27310: Localized date field formatting in admin messing up actual date

### Manual testing scenarios (*)
1. Have 2 admin users. One on locale `en_US`, one on `nl_NL`.
2. Log in on `nl_NL` admin user.
3. Create a cart price rule, no coupon code needed, just a simple cart price rule. Make sure you set the **from date** (e.g. 17 march 2020). Save the cart price rule.
4. Check if the **from date** is correct.

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by lbajsarowicz** [comment]
@tdgroot Great that you've approached the issue.
Could you cover the change with Functional tests (MFTF)?

**Review by lbajsarowicz — CHANGES_REQUESTED** [review]
@tdgroot Great that you've approached the issue.
Could you cover the change with Functional tests (MFTF)?

**Comment by tdgroot** [comment]
> @tdgroot Great that you've approached the issue.
> Could you cover the change with Functional tests (MFTF)?

Will do!

**Comment by tdgroot** [comment]
@lbajsarowicz I added an MFTF test to cover this PR. Please let me know if it's any good :smile:.

File: app/code/Magento/Ui/Test/Mftf/Test/AdminLocalizedDateElementTest.xml [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+-->
+
+<tests xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+       xsi:noNamespaceSchemaLocation="urn:magento:mftf:Test/etc/testSchema.xsd">
+    <test name="AdminLocalizedDateElementTest">
+        <annotations>
+            <stories value="Date element is localized"/>
+            <title value="Admin user should be able to create active admin user"/>
+            <description value="Admin date element should be localized."/>
+            <testCaseId value="GITHUB_27311"/>
```
Comment (lbajsarowicz) [inline app/code/Magento/Ui/Test/Mftf/Test/AdminLocalizedDateElementTest.xml]: I need to ask @okolesnyk if we should use `magento/magento2#27311` syntax instead, as we have multiple repositories and projects and the same `GITHUB_XXX` can be realted to different projects.

**Review by lbajsarowicz — CHANGES_REQUESTED** [review]
It looks like you've already worked with MFTF.
Almost perfect :clap:

We need to separate steps of preparing the test from actual testing:

**Before**
1. Generate date objects
1. Login as Main Admin
1. Create Dutch Admin
1. Logout from Main Admin

**After**
1. Login as Main Admin
1. Remove Dutch Account
1. Logout from Main Admin

**Test**
1. Login as Main Admin
1. Create price rule in en_US and make assertions
1. Logout from Main Admin
1. Login as Dutch
1. Create price rule in nl_nl and make assertions
1. Logout from dutch account

---
That would be awesome if you find the way to remove Price Rules in `After` section.

**Comment by tdgroot** [comment]
> It looks like you've already worked with MFTF.

I joined the MFTF workshop with you last year in Florence at MageTestFest!

> We need to separate steps of preparing the test from actual testing:

Will do!

> That would be awesome if you find the way to remove Price Rules in `After` section.

I was thinking about that, but somehow forgot to add that. I'll add that in!

**Comment by tdgroot** [comment]
@lbajsarowicz I added the requested changes. 

One thing I could not do was move the `generateDate` steps to before section. The variables would only be usable within the before section scope.

**Comment by lbajsarowicz** [comment]
You are right about data scope. That is fine 
Let's wait for tests result and you have :+1:

File: app/code/Magento/Ui/Test/Mftf/Test/AdminLocalizedDateElementTest.xml (hunk 1 above)
Comment (tdgroot) [reply app/code/Magento/Ui/Test/Mftf/Test/AdminLocalizedDateElementTest.xml]: I am also curious about the group and severity.

Comment (lbajsarowicz) [reply app/code/Magento/Ui/Test/Mftf/Test/AdminLocalizedDateElementTest.xml]: Don't worry about this now :-)
MFTF team have to review all of them and adjust the values there.

**Comment by tdgroot** [comment]
@magento-engcom-team what changes are requested?

**Comment by lbajsarowicz** [comment]
@tdgroot Could you fix static tests:
image
and update a branch with latest `2.4-develop`

**Comment by tdgroot** [comment]
The branch is up to date with `2.4-develop` and the static test has been fixed as well. 

Please let me know what needs to be done to get this merged.

**Comment by lbajsarowicz** [comment]
@magento run all tests

Thanks, let's have the tests green and then reviewed.

**Comment by TOIVI291** [comment]
Could you please tell me how to fix this issue @tdgroot? Because i also have this problem with dutch locale.

**Comment by lbajsarowicz** [comment]
@magento run all tests

File: app/code/Magento/Ui/Test/Mftf/Test/AdminLocalizedDateElementTest.xml (hunk 1 above)
Comment (lbajsarowicz) [reply app/code/Magento/Ui/Test/Mftf/Test/AdminLocalizedDateElementTest.xml]: @tdgroot I have response from @okolesnyk - `MC-29484`

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Review by lbajsarowicz — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @lbajsarowicz, thank you for the review.
[ENGCOM-7989](https://jira.corp.magento.com/browse/ENGCOM-7989) has been created to process this Pull Request
:eight_spoked_asterisk: @lbajsarowicz, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

File: app/code/Magento/Ui/Test/Mftf/Test/AdminLocalizedDateElementTest.xml [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
```
Comment (lbajsarowicz) [inline app/code/Magento/Ui/Test/Mftf/Test/AdminLocalizedDateElementTest.xml]:
```suggestion
            <testCaseId value="MC-29484"/>
```

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by sdzhepa** [comment]
@magento run all tests

**Comment by engcom-Alfa** [comment]
run all tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Manual testing scenario:**

1. Have 2 admin users. One on locale **en_US**, one on **nl_NL**;
2. Log in on **nl_NL** admin user;
3. Create a cart price rule, no coupon code needed, just a simple cart price rule. Make sure you set the from the date (e.g. 13 August 2020). Save the cart price rule.
4. Check if the "**from**" date is correct.

**Before:** :heavy_multiplication_x: The saved **from** date is: **20.08.2013**

Peek 2020-08-13 14-07

**After:** :heavy_check_mark: The saved **from** date is: **13.08.2020**

Peek 2020-08-13 14-00

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Comment by tdgroot** [comment]
@engcom-Alfa @engcom-Charlie @lbajsarowicz when will this PR get merged? It has been under review for over 6 months and QA has passed for the second time as well.

**Comment by ihor-sviziev** [comment]
@magento run all tests

File: app/code/Magento/Ui/Test/Mftf/Test/AdminLocalizedDateElementTest.xml [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        <annotations>
+            <stories value="Date element is localized"/>
+            <title value="Admin user should be able to create active admin user"/>
+            <description value="Admin date element should be localized."/>
+            <testCaseId value="MC-29484"/>
+            <severity value="CRITICAL"/>
+            <group value="ui"/>
+        </annotations>
+        <before>
+            <actionGroup ref="LoginAsAdmin" stepKey="LoginAsAdmin"/>
```
Comment (lbajsarowicz) [inline app/code/Magento/Ui/Test/Mftf/Test/AdminLocalizedDateElementTest.xml]:
```suggestion
            <actionGroup ref="AdminLoginActionGroup" stepKey="loginAsAdmin"/>
```

File: app/code/Magento/Ui/Test/Mftf/Test/AdminLocalizedDateElementTest.xml [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+                <argument name="role" value="genericAdminRole"/>
+                <argument name="user" value="dutchActiveAdmin"/>
+            </actionGroup>
+
+            <actionGroup ref="AdminLogoutActionGroup" stepKey="logoutMasterAdmin"/>
+        </before>
+        <after>
+            <actionGroup ref="AdminLogoutActionGroup" stepKey="logoutAdmin"/>
+
+            <actionGroup ref="LoginAsAdmin" stepKey="LoginAsAdmin"/>
```
Comment (lbajsarowicz) [inline app/code/Magento/Ui/Test/Mftf/Test/AdminLocalizedDateElementTest.xml]:
```suggestion
            <actionGroup ref="AdminLoginActionGroup" stepKey="loginAsAdmin"/>
```

File: app/code/Magento/Ui/Test/Mftf/Test/AdminLocalizedDateElementTest.xml [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        <!-- Step 0. Generate dates, this cannot be done in before section. -->
+
+        <generateDate date="now" format="m/j/Y" timezone="America/Los_Angeles" stepKey="today"/>
+        <generateDate date="now" format="d-m-Y" timezone="America/Los_Angeles" stepKey="todayNL"/>
+        <generateDate date="+1 days" format="m/j/Y" timezone="America/Los_Angeles" stepKey="tomorrow"/>
+        <generateDate date="+1 days" format="d-m-Y" timezone="America/Los_Angeles" stepKey="tomorrowNL"/>
+
+        <!-- Step 1. Log in as main admin -->
+
+        <actionGroup ref="LoginAsAdmin" stepKey="LoginAsAdmin"/>
```
Comment (lbajsarowicz) [inline app/code/Magento/Ui/Test/Mftf/Test/AdminLocalizedDateElementTest.xml]:
```suggestion
        <actionGroup ref="AdminLoginActionGroup" stepKey="loginAsAdmin"/>
```

**Review by lbajsarowicz — CHANGES_REQUESTED** [review]
🔴  World has changed since last your changes @tdgroot , we need to adjust your changes to the new world order.

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by tdgroot** [comment]
> red_circle World has changed since last your changes @tdgroot , we need to adjust your changes to the new world order.

Yes the world has actually changed a lot since I opened the PR :laughing:. Thanks for getting things going guys :).

**Comment by lbajsarowicz** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @tdgroot @lbajsarowicz,
Unfortunately tests still failing. Could you review failed tests and adjust them?

**Comment by sidolov** [comment]
@tdgroot  I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for the collaboration!

**Comment by tdgroot** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE

**Comment by tdgroot** [comment]
@sidolov Can you reopen please? I want to finish this PR!

**Comment by ihor-sviziev** [comment]
@tdgroot sure, just re-opened!

**Comment by tdgroot** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE

**Comment by tdgroot** [comment]
@ihor-sviziev @lbajsarowicz I see the following error:
`select option "#page_tabs_main_section_content select[name='interface_locale']","nl_NL" `

The `nl_NL` locale is not deployed in the Magento test system, so the `nl_NL` locale cannot be selected when creating an admin user.

How do we go about that? To me it seems very important to be able to test the application on other locales as well.

**Comment by ihor-sviziev** [comment]
@lbajsarowicz  could you help with the issue in MFTF?

**Comment by tdgroot** [comment]
@ihor-sviziev I updated the branch to resolve merge conflicts. Could you let me know what the status is of this issue?

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @tdgroot,
Please review and fix failing tests

**Comment by barryvdh** [comment]
I can confirm this issue also occurs for the special_from/to dates on the catalog edit page. Example:

Special price from 2021-02-01 - 2021-03-01, shows in the English interface:
image

Change interface to dutch, dat now shows like this:
image

Without changing anything, just pressing the Save button, now changes the date from the product, also in English:
image

Bonus: if you open it again in NL it changes again!
image

So basically the admin is unusable on Dutch date format.

**Comment by barryvdh** [comment]
@lbajsarowicz @ihor-sviziev Could you help with the tests? We have to disable the NL locale in all our back-ends because it changes all dates upon saving, which is very confusing for our clients.

**Comment by barryvdh** [comment]
Is this fixed by https://github.com/magento/magento2/pull/31047 or just partially?

**Comment by PascalBrouwers** [comment]
Confirmed that this PR works. Thank you tdgroot.
What I don't get is why this STILL isn't fixed since the release of Magento 2.0.
Magento maybe help with the tests if you're so adement on having this required. People stop contributing because of the extreme bad experience getting a PR merged.

**Comment by ihor-sviziev** [comment]
@Den4ik, maybe you could review and help with tests?

**Comment by Den4ik** [comment]
This is related to https://github.com/magento/magento2/pull/33591
I'm trying implement correct ICU convert for all date UI components

**Comment by PascalBrouwers** [comment]
Looking at https://github.com/magento/magento2/pull/31047. This one works better and is less code change.

**Comment by Den4ik** [comment]
But in fact it's masking of talk problem. Look at `normalizeDate` function. This method replacing year and day placeholders because correct converting ICU dates format to jq calendar and momentjs is not working

**Comment by joshuaswarren** [comment]
@tdgroot - thank you for the effort you put into this pull request! It's been out here a while - it's actually one of the oldest pull requests still open. In your opinion, is this still an issue? Pascal's comment from 2021 makes it seem like the merged pull request https://github.com/magento/magento2/pull/31047 may also have addressed this problem. And Igor pointed out that there's failed tests that need to be addressed. 

What do you think, is there a need to put the  additional effort into this PR to get it merged, or is it out of date now and could be closed?

**Comment by tdgroot** [comment]
Hi @joshuaswarren,

Yeah it's been a while 😅. I'm not actively developing Magento projects anymore, so I haven't been in the in the loop on this one.

It could very well be the case that the problem has been solved, but I know there were multiple problems that required various fixes. It might be good to ask @barryvdh or @PascalBrouwers if this is still a problem.

Anyway, last time I checked, the tests were failing because I introduced something that the test suite didn't support: multiple admin locales. So the tests change the admin user locale to `nl_NL`< but that locale isn't available in the test runner, because it only seems to deploy `en_US`.

**Comment by engcom-Echo** [comment]
@tdgroot 
Thank you for your contribution and collaboration.
The issue seems to fixed under this [commit](https://github.com/magento/magento2/pull/31047).

Hence Closing this PR.
If you are still facing the Issue please feel free to reopen the issue.
